libs/s25main/notifications/NotificationManager_impl.h

```diff
@@ -58,62 +40,94 @@ struct NotificationManager::NoteCallbackBase
-    for(auto& itSubscribers : noteId2Subscriber)
-        for(auto& itCallback : itSubscribers.second)
-            static_cast<NoteCallbackBase*>(itCallback)->SetUnsubscribed();
+    for(const auto& subscribers : noteId2Subscriber)
+    {
+        RTTR_Assert(!subscribers.second.isPublishing);
+        for(void* callback : subscribers.second.callbacks)
+        {
+            if(callback)
+                static_cast<NoteCallbackBase*>(callback)->SetUnsubscribed();
+        }
+    }
 }
 
-inline void NotificationManager::unsubscribe(Subscription& subscription)
+inline void NotificationManager::unsubscribe(Subscription& subscription) noexcept
 {
     // We can simply call reset as this calls the deleter which releases the subscription
     subscription.reset();
 }
 
 template<class T_Note>
-Subscription NotificationManager::subscribe(std::function<void(T_Note)> callback)
+Subscription NotificationManager::subscribe(std::function<void(const T_Note&)> callback) noexcept
 {
-    if(isPublishing)
-        throw std::runtime_error("Cannot subscribe during publishing of messages");
-    auto* subscriber = new NoteCallback<T_Note>(callback);
-    noteId2Subscriber[T_Note::getNoteId()].push_back(subscriber);
-    return Subscription(subscriber, CallbackUnregistrar<T_Note>(*this));
+    auto* subscriber = new NoteCallback<T_Note>(std::move(callback));
+    noteId2Subscriber[T_Note::getNoteId()].callbacks.push_back(subscriber);
+
+    return Subscription(subscriber, [this](void* subscription) {
+        RTTR_Assert(subscription); // As we use this in a shared_ptr, this can never be nullptr
+        auto* callback = static_cast<NoteCallback<T_Note>*>(subscription);
+        if(callback->IsSubscribed())
+            this->unsubscribe<T_Note>(callback);
+        delete callback;
+    });
 }
 
 template<class T_Note>
-void NotificationManager::unsubscribe(NoteCallback<T_Note>* callback)
+void NotificationManager::unsubscribe(NoteCallback<T_Note>* callback) noexcept
 {
-    if(!callback->IsSubscribed())
-        return;
-    if(isPublishing)
-        throw std::runtime_error("Cannot unsubscribe during publishing of messages");
-    CallbackList& callbacks = noteId2Subscriber[T_Note::getNoteId()];
+    RTTR_Assert(callback->IsSubscribed());
+    Subscribers& subs = noteId2Subscriber[T_Note::getNoteId()];
+    CallbackList& callbacks = subs.callbacks;
     auto itEl = std::find(callbacks.begin(), callbacks.end(), callback);
-    if(itEl != callbacks.end())
-        callbacks.erase(itEl);
+    RTTR_Assert(itEl != callbacks.end());
+    // We can't modify the list while iterating over it
+    if(subs.isPublishing)
+        *itEl = nullptr;
     else
-        RTTR_Assert(false);
+        callbacks.erase(itEl);
     // Set only the flag. Actual deletion of the callback must be handled by the shared_ptr deleter
     callback->SetUnsubscribed();
 }
 
 template<class T_Note>
 void NotificationManager::publish(const T_Note& notification)
```

Comment:
normally this is named `notify` ?

Reply:
Not in this case. You **notify** a subscriber but you **publish** a notification. In this case it is a "broadcast to whoever listens" so "publish" makes more sense. See e.g. the callsite: `manager.publish(MyNotification{foo});`